<a href="https://colab.research.google.com/github/Doclikam/Machine-Learning-2/blob/main/week3_deep%20learning/dlpractice/buildingCNN.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
#importing livaries
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers
from tensorflow.keras.datasets import cifar10

In [ ]:
(x_train,y_train), (x_test, y_test)=cifar10.load_data()
x_train=x_train.astype('float32')/255.0
x_test=x_test.astype('float32')/255.0

170498071/170498071 ━━━━━━━━━━━━━━━━━━━━ 3s 0us/step


In [ ]:
x_train.shape

(50000, 32, 32, 3)

We are not going to reshape it since we are us cnn

In [ ]:
model=tf.keras.Sequential(
    [keras.Input(shape=(32,32,3)),
    layers.Conv2D(32, 3, padding='valid', activation='relu'),
     layers.MaxPooling2D(pool_size=(2,2)),
     layers.Conv2D(64, 3, activation='relu'),
     layers.MaxPooling2D(),
     layers.Conv2D(128,3, activation='relu'),
     layers.MaxPooling2D(),
     layers.Flatten(),
     layers.Dense(64,activation='relu'),
     layers.Dense(10)
     ]
)
print(model.summary())
model.compile(loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
              optimizer=tf.keras.optimizers.Adam(learning_rate=3e-4), metrics=['accuracy'])
model.fit(x_train,y_train, batch_size=64, epochs=10, verbose=2)
model.evalaute(x_test, y_test, batch_size=64, verbose=2)


Model: "sequential_2"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ conv2d_6 (Conv2D)                    │ (None, 30, 30, 32)          │             896 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ max_pooling2d_4 (MaxPooling2D)       │ (None, 15, 15, 32)          │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ conv2d_7 (Conv2D)                    │ (None, 13, 13, 64)          │          18,496 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ max_pooling2d_5 (MaxPooling2D)       │ (None, 6, 6, 64)            │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ conv2d_8 (Conv2D)                    │ (None, 4, 4, 128)           │          73,856 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ max_pooling2d_6 (MaxPooling2D)       │ (None, 2, 2, 128)           │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ flatten (Flatten)                    │ (None, 512)                 │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense (Dense)                        │ (None, 64)                  │          32,832 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_1 (Dense)                      │ (None, 10)                  │             650 │
└──────────────────────────────────────┴─────────────────────────────┴─────────────────┘

 Total params: 126,730 (495.04 KB)

 Trainable params: 126,730 (495.04 KB)

 Non-trainable params: 0 (0.00 B)

None
Epoch 1/10
782/782 - 70s - 89ms/step - accuracy: 0.3640 - loss: 1.7446
Epoch 2/10
782/782 - 79s - 100ms/step - accuracy: 0.4989 - loss: 1.3947
Epoch 3/10
782/782 - 83s - 106ms/step - accuracy: 0.5485 - loss: 1.2756
Epoch 4/10
782/782 - 83s - 106ms/step - accuracy: 0.5828 - loss: 1.1886
Epoch 5/10
782/782 - 84s - 107ms/step - accuracy: 0.6115 - loss: 1.1131
Epoch 6/10
782/782 - 79s - 102ms/step - accuracy: 0.6335 - loss: 1.0590
Epoch 7/10
782/782 - 64s - 82ms/step - accuracy: 0.6505 - loss: 1.0124
Epoch 8/10
782/782 - 85s - 109ms/step - accuracy: 0.6665 - loss: 0.9699
Epoch 9/10
782/782 - 66s - 84ms/step - accuracy: 0.6806 - loss: 0.9292
Epoch 10/10
782/782 - 82s - 104ms/step - accuracy: 0.6920 - loss: 0.8959


AttributeError: 'Sequential' object has no attribute 'evalaute'


 [keras.Input(shape=(32,32,3(3 channels)),
layers.Conv2D(32( this argument indicates the number of channels we want the convulution layer to output), (3,3)(kernel size)]

   parameters:

Let's say that we have a dense layer with an output of 8 and you connect it to 4 neurons. The number of weights is 8*4=32, biases=4


Whe it comes to CNN thr learnable parameters are in the filters/kernels:
For a CNN filter size k*k , c input channels, and f filters. The filter is a tensor of shape (k *k *c). Total weight parameters (k * k * c * f). There is one bias per filter so f.

Ie: filter size (3*3), input channels 3, number of filters 32 , weights=3*3*3*32=864    biases =32           32+864=896 parameters

# Batch Normalization


it increases the speed of training and decreases the importance of initial weights, and plays a small role in regularizing the model.

# Dimensionality Reduction

feature selection:   wrappers, filters, embedded

feature extraction: pca


In [ ]:
#functional model
def my_model():
  inputs=keras.Input(shape=(32,32,3))
  x=layers.Conv2D(32,3 )(inputs)
  x=layers.BatchNormalization()(x)
  x=keras.activations.relu(x)
  x=layers.MaxPooling2D()(x)
  x=layers.Conv2D(64,5, padding='same')(x)
  x=layers.BatchNormalization()(x)
  x=keras.activations.relu(x)
  x=layers.Conv2D(128, 3)(x)
  x=layers.BatchNormalization()(x)
  x=keras.activations.relu(x)
  x=layers.Flatten()(x)
  x=layers.Dense(64, activation='relu')(x)
  outputs=layers.Dense(10)(x)
  model=keras.Model(inputs=inputs, outputs=outputs)
  return model


model=my_model()

model.compile(loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
              optimizer=tf.keras.optimizers.Adam(learning_rate=3e-4), metrics=['accuracy'])
model.fit(x_train, y_train, batch_size=64, epochs=10, verbose=2)
model.evaluate(x_test, y_test, batch_size=64, verbose=2)





Epoch 1/10
782/782 - 278s - 356ms/step - accuracy: 0.4630 - loss: 1.4643
Epoch 2/10
782/782 - 317s - 406ms/step - accuracy: 0.6421 - loss: 1.0156
Epoch 3/10
782/782 - 325s - 416ms/step - accuracy: 0.7065 - loss: 0.8330
Epoch 4/10
782/782 - 318s - 407ms/step - accuracy: 0.7457 - loss: 0.7284
Epoch 5/10
782/782 - 319s - 408ms/step - accuracy: 0.7742 - loss: 0.6452
Epoch 6/10
782/782 - 328s - 419ms/step - accuracy: 0.7995 - loss: 0.5744
Epoch 7/10
782/782 - 317s - 405ms/step - accuracy: 0.8212 - loss: 0.5156
Epoch 8/10
782/782 - 268s - 343ms/step - accuracy: 0.8416 - loss: 0.4532
Epoch 9/10
782/782 - 321s - 411ms/step - accuracy: 0.8642 - loss: 0.3967
Epoch 10/10
782/782 - 325s - 416ms/step - accuracy: 0.8800 - loss: 0.3516
157/157 - 14s - 88ms/step - accuracy: 0.6649 - loss: 1.1483


[1.14828622341156, 0.664900004863739]

**Max pooling**


Max pooling is commonly used for:  dimnsionality reduction, and as a way of down sampling. This helps to reduce the computtional load and ultimately overfitting.

# **Regularization**

drop out


d3=np.random.rand(a3.shape[0], a3.shape[1]) less than keep_prob
a3=np.multiply(a3,d3)


l2:
l(O)=loss_function+regularization term

In [ ]:
def my_model():
    inputs=keras.Inputs(shape(32,32,3))
    x=layers.Conv2D(32,3, padding='same', kernel_regularizer=rugularizers.l2(0.01,))(inputs)
    x=layers.BatchNormalization()(x)
    x=keras.activation.relu(x)
    x=layers.Maxpooling2D()(x)
    x=layers.Conv2D(64,5,)(inputs)
    x=layers.BatchNormalization()(x)
    x=keras.activation.relu(x)
    x=layers.Conv2D(128,3,)(inputs)
    x=layers.BatchNormalization()(x)
    x=keras.activation.relu(x)
    x=layers.Flatten()(x)
    x=layers.Dense(64, activation='relu', kernel_regularizer=rugularizers.l2(0.01,))(x)
    x=layers.dropout(0.5)(x)
    ouputs=layers.Dense(10)(x)
    model=keras.Model(inputs=inputs, outputs=outputs)
    return model


  model=my_model()
  model.compile(loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
              optimizer=tf.keras.optimizers.Adam(learning_rate=3e-4), metrics=['accuracy'])
  model.fit(x_train, y_train, batch_size=64, epochs=150, verbose=2)
  model.evaluate(x_test, y_test, batch_size=64, verbose=2)

long short Term Memory(lstm)

This is a type of RNN that solves the vanishing and exploding gradient problem.    

in this case it uses a path for long termmemory and a path for shprt term memo
ries to make predictions. It adds an internal state to the RNN

forget gate-------->input gate---------->output gate
(0,1)

uses: machine translation,  chat bot



In [ ]:
from tensorflow.keras.datasets import mnist

In [ ]:
(x_train, y_train), (x_test, y_test)=mnist.load_data()
x_train=x_train.astype('float32')/255.0
x_test=x_test.astype('float32')/255.0

In [ ]:
model=keras.Sequential()
model.add(keras.Input(shape=(None,28)))
model.add(layers.SimpleRNN(512, return_sequences=True, activation='relu'))
model.add(layers.SimpleRNN(512, activation='relu'))
model.add(layers.Dense(10))

print(model.summary())



Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ simple_rnn (SimpleRNN)               │ (None, None, 512)           │         276,992 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ simple_rnn_1 (SimpleRNN)             │ (None, 512)                 │         524,800 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense (Dense)                        │ (None, 10)                  │           5,130 │
└──────────────────────────────────────┴─────────────────────────────┴─────────────────┘

 Total params: 806,922 (3.08 MB)

 Trainable params: 806,922 (3.08 MB)

 Non-trainable params: 0 (0.00 B)

None


In [ ]:
model.compile(
    loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
              optimizer=tf.keras.optimizers.Adam(learning_rate=0.001),
    metrics=['accuracy'])
model.fit(x_train, y_train, batch_size=64, epochs=10, verbose= 2)
model.evaluate(x_test, y_test, batch_size=64, verbose=2)

Epoch 1/10
938/938 - 244s - 260ms/step - accuracy: 0.9817 - loss: 0.0697
Epoch 2/10
938/938 - 258s - 275ms/step - accuracy: 0.9847 - loss: 0.0579
Epoch 3/10
938/938 - 262s - 279ms/step - accuracy: 0.9823 - loss: 0.0650
Epoch 4/10
938/938 - 266s - 284ms/step - accuracy: 0.9834 - loss: 0.0614
Epoch 5/10
938/938 - 260s - 277ms/step - accuracy: 0.9858 - loss: 0.0498
Epoch 6/10
938/938 - 246s - 262ms/step - accuracy: 0.9857 - loss: 0.0521
Epoch 7/10
938/938 - 242s - 258ms/step - accuracy: 0.9859 - loss: 0.0509
Epoch 8/10
938/938 - 257s - 274ms/step - accuracy: 0.9868 - loss: 0.0493
Epoch 9/10
938/938 - 264s - 281ms/step - accuracy: 0.9861 - loss: 0.0513
Epoch 10/10
938/938 - 260s - 277ms/step - accuracy: 0.9875 - loss: 0.0453
157/157 - 12s - 76ms/step - accuracy: 0.9804 - loss: 0.0801


[0.08009346574544907, 0.980400025844574]

In [ ]:
model=keras.Sequential()
model.add(keras.Input(shape=(None,28)))
model.add(layers.GRU(256, return_sequences=True, activation='tanh'))
model.add(layers.GRU(256, activation='tanh'))
model.add(layers.Dense(10))

print(model.summary())
model.compile(
    loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
              optimizer=tf.keras.optimizers.Adam(learning_rate=0.001),
    metrics=['accuracy'])
model.fit(x_train, y_train, batch_size=64, epochs=10, verbose= 2)
model.evaluate(x_test, y_test, batch_size=64, verbose=2)

In [ ]:
model=keras.Sequential()
model.add(keras.Input(shape=(None,28)))
model.add(
    layers.Bidirectional(
        layers.LSTM(256, return_sequences=True, activation='tanh')))
model.add(
    layers.Bidirectional(
        layers.LSTM(256, activation='tanh')))
model.add(layers.Dense(10))

print(model.summary())
model.compile(
    loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
              optimizer=tf.keras.optimizers.Adam(learning_rate=0.001),
    metrics=['accuracy'])
model.fit(x_train, y_train, batch_size=64, epochs=10, verbose= 2)
model.evaluate(x_test, y_test, batch_size=64, verbose=2)

# **Functional API**

In [ ]:
import pandas as pd
import matplotlib.pyplot as plt


In [ ]:
from tensorflow.keras import layers, regularizers

Remember sequential maps one input to one output, in cases where we have more than one inpu then we can use sequential.

In [ ]:
def model():
  input=keras.Input(shape=(64,64,1))
  x=layers.Conv2D(
      filters= 32,
      kernel_size=3,
      padding ='same',
      kernel_rugularizer=

  )input


In [ ]:
data=pandas

# **Model SUbclassing**

In [ ]:
(x_train, y_train), (x_test, y_test)=mnist.load_data()
x_train=x_train.reshape(-1, 28,28,1).astype('float32')/255.0
x_test=x_test.reshape(-1,28,28,1).astype('float32')/255.0

In [ ]:
class CNNBlock(layers.Layers):
  def __init__(self, out_channels, kernel_size=3):
    super(CNNBlock, self).
